### PR TITLE
Environment variable documentation

### PR DESCRIPTION
**Pull request by tpinney**
### Context

Added documentation to explain how to add environment variables in local development and Azure hosted environments.

### Changes proposed in this pull request

- Added environment variables documentation.
- Added table of contents to readme to aid readability.

### Guidance to review

README.md is only file that has been changed. Resultant output of change can be found here:
https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/tree/environment_variables_documentation

### Link to Trello card

N/A

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -62,15 +79,71 @@ check` before proceeding to debug using `bin/webpack`.
...
+
+#### <a name="documentation-env-vars-docker-compose"></a>Docker Compose
+
+For docker compose to make the necessary environment variables available in the container at run time they must be declared in the relevant docker-compose.yaml file, of which there are three.
+
+* [docker-compose.yml](./docker-compose.yml) - Variables that are required for local dev and the docker image build phase in azure only should be defined the *environment* section in this file. This will in general only be for the database.
+* [docker-compose.override.yml](./docker-compose.override.yml) - This file is used exclusively for local development and imports variables from the .env file so no further changes are required here.
+* [docker-compose.azure.yml](./docker-compose.azure.yml) - This file is exclusively used in the Azure devops pipeline. Any environment variables required in the Azure build/deployment need to be delcared in the *environment* section of this file. **You should only declare the environment variable here, not its value.** The only exception to this is where we need to map variables due to the use of protected variable names, e.g. Rails SECRET_KEY_BASE.
```
Comment (duncanjbrown) [inline README.md]: Typo: `delcared`

File: README.md [hunk 2]
```diff
@@ -62,15 +79,71 @@ check` before proceeding to debug using `bin/webpack`.
...
+`VARIABLE_NAME=Value`
+
+The [.env.example](./.env.example) file contains the essential environment variables that must exist for local development builds to succeed.
+
+#### <a name="documentation-env-vars-docker-compose"></a>Docker Compose
+
+For docker compose to make the necessary environment variables available in the container at run time they must be declared in the relevant docker-compose.yaml file, of which there are three.
+
+* [docker-compose.yml](./docker-compose.yml) - Variables that are required for local dev and the docker image build phase in azure only should be defined the *environment* section in this file. This will in general only be for the database.
+* [docker-compose.override.yml](./docker-compose.override.yml) - This file is used exclusively for local development and imports variables from the .env file so no further changes are required here.
```
Comment (duncanjbrown) [inline README.md]: `docker-compose`'s behaviour with `.env` is a bit funky. From the docs I think it uses `.env` to resolve environment variables inside the `docker-compose.yml` file itself (https://docs.docker.com/compose/environment-variables/#the-env-file). It's a happy accident that `.env` is copied inside the container where Rails can pick up its values. Could we just say "this file is used exclusively for local development"?

File: README.md [hunk 3]
```diff
@@ -62,15 +79,71 @@ check` before proceeding to debug using `bin/webpack`.
...
+1. In the [azure/template.json](./azure/template.json) file you need to make the following additions:
+   1. Duplicate this block of code for your new variable in the *parameters* section at the start of the file.
+      ```"var_name": {
+       "type": "string",
+        "metadata": {
+          "description": "Describe your variable here."
+        }
+      }
+      ```
+   1. Around line 180 duplicate this block of into the *appSericeAppSettings* section.
```
Comment (duncanjbrown) [inline README.md]: Typo: `appSerice`

File: README.md [hunk 4]
```diff
@@ -62,15 +79,71 @@ check` before proceeding to debug using `bin/webpack`.
...
+
+These steps describe the process for making environment variables available to the the Azure DevOps pipeline.
+
+1. Declare the desired variable in the appropriate "variable group" in the Library section of the Azure DevOps site (https://dfe-ssp.visualstudio.com/Become-A-Teacher/_library?itemType=VariableGroups). All variable groups related to apply are suffixed as such and there is a variable group per deployment environment.
+1. In the [azure-pipelines.yml](./azure-pipelines.yml) file there are several changes to be made:
+   1. For each "make" command script step, add your environment variable to the *env* section in the format `ENV_VAR_NAME: $(var_name)` where **ENV_VAR_NAME** is the environment variable name as it should appear in the docker container and **var_name** is the name of the variable defined in the Azure DevOps variable group.
+   1. For each 'deployment stage' you must add your variable to the template *parameters* section in the format `var_name: '$(var_name)'`.
+1. In the [azure-pipelines-deploy-template.yaml](./azure-pipelines-deploy-template.yaml) file you need to make the following additions:
```
Comment (duncanjbrown) [inline README.md]: Broken link - should be `.yml`

File: README.md [hunk 5]
```diff
@@ -62,15 +79,71 @@ check` before proceeding to debug using `bin/webpack`.
...
+   1. Add your variable to the *parameters* section at the start of the file using the name of the variable as it appears in the variable group.
+   1. In the Azure Resource Group deployment task *overrideParameters* section (around line 50) add your variable in the format `-var_name: "${{parameters.var_name}}"`
+1. In the [azure/template.json](./azure/template.json) file you need to make the following additions:
+   1. Duplicate this block of code for your new variable in the *parameters* section at the start of the file.
+      ```"var_name": {
```
Comment (duncanjbrown) [inline README.md]: The first line of this and the following code block are cut off when rendered — maybe it wants a line break after the backticks?

File: README.md [hunk 6]
```diff
@@ -62,15 +79,71 @@ check` before proceeding to debug using `bin/webpack`.
...
+1. In the [azure-pipelines.yml](./azure-pipelines.yml) file there are several changes to be made:
+   1. For each "make" command script step, add your environment variable to the *env* section in the format `ENV_VAR_NAME: $(var_name)` where **ENV_VAR_NAME** is the environment variable name as it should appear in the docker container and **var_name** is the name of the variable defined in the Azure DevOps variable group.
+   1. For each 'deployment stage' you must add your variable to the template *parameters* section in the format `var_name: '$(var_name)'`.
+1. In the [azure-pipelines-deploy-template.yaml](./azure-pipelines-deploy-template.yaml) file you need to make the following additions:
+   1. Add your variable to the *parameters* section at the start of the file using the name of the variable as it appears in the variable group.
+   1. In the Azure Resource Group deployment task *overrideParameters* section (around line 50) add your variable in the format `-var_name: "${{parameters.var_name}}"`
```
Comment (duncanjbrown) [inline README.md]: The line reference here feels a bit risky as it might go out of date

File: README.md [hunk 7]
```diff
@@ -62,15 +79,71 @@ check` before proceeding to debug using `bin/webpack`.
...
+
+#### <a name="documentation-env-vars-azure-devops"></a>Azure Hosting (DevOps pipeline)
+
+These steps describe the process for making environment variables available to the the Azure DevOps pipeline.
+
+1. Declare the desired variable in the appropriate "variable group" in the Library section of the Azure DevOps site (https://dfe-ssp.visualstudio.com/Become-A-Teacher/_library?itemType=VariableGroups). All variable groups related to apply are suffixed as such and there is a variable group per deployment environment.
+1. In the [azure-pipelines.yml](./azure-pipelines.yml) file there are several changes to be made:
+   1. For each "make" command script step, add your environment variable to the *env* section in the format `ENV_VAR_NAME: $(var_name)` where **ENV_VAR_NAME** is the environment variable name as it should appear in the docker container and **var_name** is the name of the variable defined in the Azure DevOps variable group.
```
Comment (duncanjbrown) [inline README.md]: Should we use `varName` here and below instead of `var_name`, as that's the convention so far?

File: README.md [hunk 8]
```diff
@@ -5,27 +5,44 @@
...
+* [Prerequisites for development](#dev-prerequisites)
+* [Setting up the development environment](#dev-env-setup)
+* [Docker workflow](#docker-workflow)
+* [Webpacker](#webpacker)
+  * [Debugging Webpacker](#webpacker-debug)
+* [Documentation](#documentation)
+  * [Nomenclature](#documentation-nomenclature)
+  * [Domain Model](#documentation-domain-model)
+  * [Environment Variables](#documentation-env-vars)
+    * [Dockerfile](#documentation-env-vars-dockerfile)
```
Comment (duncanjbrown) [inline README.md]: This table of contents is very useful. I find the third level of nesting makes it less easy to scan, though — do you agree? Happy to leave it in if not

Comment (tpinney) [reply README.md]: Happy to remove this if others agree. I don't particularly have a view either way.

File: README.md (hunk 4 above)
Comment (tpinney) [inline README.md]:
```suggestion
1. In the [azure-pipelines-deploy-template.yaml](./azure-pipelines-deploy-template.yml) file you need to make the following additions:
```

File: README.md (hunk 3 above)
Comment (tpinney) [inline README.md]:
```suggestion
   1. Around line 180 duplicate this block of into the *appServiceAppSettings* section.
```

File: README.md (hunk 2 above)
Comment (tpinney) [reply README.md]:
```suggestion
* [docker-compose.override.yml](./docker-compose.override.yml) - This file is used exclusively for local development. No further changes are required here.
```

File: README.md (hunk 6 above)
Comment (tpinney) [inline README.md]:
```suggestion
   1. In the Azure Resource Group deployment task *overrideParameters* section add your variable in the format `-var_name: "${{parameters.var_name}}"`
```

File: README.md (hunk 7 above)
Comment (tpinney) [reply README.md]: Agreed. Implemented in most recent commit.

File: README.md (hunk 5 above)
Comment (tpinney) [reply README.md]: Resolved since last review.

File: README.md (hunk 4 above)
Comment (tpinney) [reply README.md]: Corrected.

File: README.md (hunk 3 above)
Comment (tpinney) [reply README.md]: Corrected.

**Comment by tpinney** [comment]
@duncanjbrown please can you review again. I believe I have addressed all your comments.

**Pull request merged**